File: Algorithms/Remove Nth Node From End of List.js
// https://leetcode.com/submissions/detail/81325338/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 *
 * Given 1->2->3->4->5, n = 2, return 1->2->3->5
 */
var removeNthFromEnd = function(head, n) {
  var arr = [], len
  if (!head.next)
    return []
  while (head) {
    arr.push(head)
    head = head.next
  }
  len = arr.length
  
  if (n == len)
    return arr[1]
  arr[len - n - 1].next = arr[len - n].next
  return arr[0]
};

File: 019-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	st := []*ListNode{}
	for n := head; n != nil; {
		st = append(st, n)
		n = n.Next
	}

	idx := len(st) - n - 1
	if idx < 0 {
		return head.Next
	}
	node := st[idx]
	node.Next = node.Next.Next
	return head
}
